Pass Bundle with additional information to the drop in service

BenedictP opened the pull request:
<!-- 🎉 Thank you for submitting a pull request! 🎉  -->

## Summary
When making a payment request to our backend, I need additional data in the `onPaymentsCallRequested` and `onDetailsCallRequested` method. I didn't want to use shared prefs (what you are doing in the example app) or something else (e.g. inject a "data holder" class). So I added this functionality into my fork and now I want to show you what I did.
So you can merge this if you think it could be helpful or close it if this is not useful.


## Tested scenarios
<!-- Description of tested scenarios -->


**Fixed issue**: 
- https://github.com/Adyen/adyen-android/issues/180

caiofaustino commented:
Thanks for your PR! And sorry for taking so long to get to review it. This looks good! I only have 2 small comments:

1 - When I initially thought about this feature, I thought of adding the extra bundle as part of the `DropInConfiguration` instead of a new parameter in `startPayment`. This way it can be just an optional field there, we don't break the public API, and the setup to make it serializable is already there. What do you think?

2 - We have CI check for linting and code checks, this is failing for a few small things.
https://github.com/Adyen/adyen-android/runs/4964481801?check_suite_focus=true#step:5:1944
You can run this yourself before pushing by running `./gradlew check`.

BenedictP commented:
No problem, no urgency here.
1. I changed it but I still broke public API when I added the Bundle to the `DropInServiceInterface` 
2. Now everything is fine. Thanks for the hint :)

jreij commented:
Hey @BenedictP, thanks for the PR, great effort and very useful feature!

I was reading through the implementation and I have some comments about it:
-  As discussed earlier, the current solution breaks 4 public methods which means we cannot include it in a minor release (4.x).
- These 4 methods will now expose an `additionalDataForDropInService: Bundle?` parameter that can most definitely confuse anyone who does not want to use this bundle or has no idea about it, which is the majority of our merchants.

However I think we can make a small adjustment to improve it:
- What if we just add a `getAdditionalData()` method, inside `DropInService`, which you can call at any time you want to retrieve the bundle?
- In addition to not breaking any existing methods, it will also not be exposed to anyone who does not need it, avoiding possible confusion.
- And since the bundle will not change anyway and is not affected at all by the flow itself, there is no problem having it in one place, accessible at any time you might need it.

As for the technical details of the implementation, instead of using `DropInServiceInterface` we can pass the `additionalDataForDropInService: Bundle?` parameter to the `bindService` method and retrieve it in `onBind`:


```kotlin

    companion object {
        private const val INTENT_EXTRA_ADDITIONAL_DATA = "ADDITIONAL_DATA"

        internal fun bindService(
            context: Context,
            connection: ServiceConnection,
            merchantService: ComponentName,
            additionalData: Bundle?
        ): Boolean {
            Logger.d(TAG, "bindService - ${context::class.simpleName}")
            val intent = Intent().apply {
                component = merchantService
                putExtra(INTENT_EXTRA_ADDITIONAL_DATA, additionalData)
            }
            return context.bindService(intent, connection, Context.BIND_AUTO_CREATE)
        }
    }

    private var additionalData: Bundle? = null

    override fun onBind(intent: Intent?): IBinder {
        Logger.d(TAG, "onBind")
        if (intent?.hasExtra(INTENT_EXTRA_ADDITIONAL_DATA) == true) {
            additionalData = intent.getBundleExtra(INTENT_EXTRA_ADDITIONAL_DATA)
        }
        return binder
    }

    protected fun getAdditionalData(): Bundle? {
        return additionalData
    }
```

Let me know your thoughts 🙂

jreij commented:
@BenedictP I will copy your branch (since I can't take over it) and update it a bit in a new PR, if you don't mind. Thanks again for the contribution.

BenedictP commented:
No problem. I didn't have the time to work on it lately so I'm quite happy that you are picking this up :)
